File: tree/binary_tree/0144_binary_tree_preorder_traversal.js
/*
Given the root of a binary tree, return the preorder traversal of its nodes' values.

Definition for a binary tree node:
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}
*/

// https://leetcode.com/problems/binary-tree-preorder-traversal/description/

// recursive solution
var preorderTraversal = function (root) {
  if (!root) return [];
  let result = [];
  result.push(root.val);
  if (root.left) result = result.concat(preorderTraversal(root.left));
  if (root.right) result = result.concat(preorderTraversal(root.right));
  return result;
};

// modularized recursive solution
var preorderTraversal = function (root) {
  function traverse(node, result) {
    if (!node) return;
    result.push(node.val);
    traverse(node.left, result);
    traverse(node.right, result);
  }
  let result = [];
  traverse(root, result);
  return result;
};

// iterative solution + stack
var preorderTraversal = function (root) {
  if (!root) return [];
  let result = [];
  let stack = [root];
  while (stack.length) {
    let node = stack.pop();
    result.push(node.val);
    // first in last out, thus push right first
    if (node.right) stack.push(node.right);
    if (node.left) stack.push(node.left);
  }
  return result;
};

File: tree/traversal/0144_binary_tree_preorder_traversal.py
# Given the root of a binary tree, 
# return the preorder traversal of its nodes' values.

# 前序遍历 pre-order NLR
# https://leetcode.com/problems/binary-tree-preorder-traversal/

# recursive
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        def preorder(root):
            # base case
            if not root:
                return
            # recursive case
            res.append(root.val)
            preorder(root.left)
            preorder(root.right)
        preorder(root)
        return res

# iterative + stack        
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:                       
            return []            
        res, stack = [], [root]
        while stack:
            # 弹出根节点                        
            node = stack.pop()
            # 访问根节点              
            res.append(node.val)
            # 右子树入栈            
            if node.right:
                stack.append(node.right)
            # 左子树入栈    
            if node.left:
                stack.append(node.left)     
        return res